QueryInterceptor.visitClearCommand blocks

### Describe the bug

```
java.lang.AssertionError: Blocking call! jdk.internal.misc.Unsafe#park on thread Thread[#4471,non-blocking-thread-CacheResourceV2Test-NodeA-p912-t2,5,ISPN-non-blocking-group]
	at org.infinispan.util.CoreTestBlockHoundIntegration.lambda$applyTo$0(CoreTestBlockHoundIntegration.java:51)
	at reactor.blockhound.BlockHound$Builder.lambda$install$8(BlockHound.java:488)
	at reactor.blockhound.BlockHoundRuntime.checkBlocking(BlockHoundRuntime.java:89)
	at java.base/jdk.internal.misc.Unsafe.park(Unsafe.java)
	at java.base/java.util.concurrent.locks.LockSupport.park(LockSupport.java:221)
	at java.base/java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1864)
	at java.base/java.util.concurrent.ForkJoinPool.unmanagedBlock(ForkJoinPool.java:3778)
	at java.base/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3723)
	at java.base/java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1898)
	at java.base/java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2117)
	at org.hibernate.search.util.common.impl.Futures.unwrappedExceptionJoin(Futures.java:192)
	at org.infinispan.search.mapper.scope.impl.SearchWorkspaceImpl.purge(SearchWorkspaceImpl.java:19)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at org.infinispan.query.backend.QueryInterceptor.purgeAllIndexes(QueryInterceptor.java:351)
	at org.infinispan.query.backend.QueryInterceptor.processClearCommand(QueryInterceptor.java:473)
	at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNextThenAccept(BaseAsyncInterceptor.java:109)
	at org.infinispan.query.backend.QueryInterceptor.visitClearCommand(QueryInterceptor.java:299)
	at org.infinispan.commands.write.ClearCommand.acceptVisitor(ClearCommand.java:33)
	at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNext(BaseAsyncInterceptor.java:58)
	at org.infinispan.interceptors.locking.AbstractLockingInterceptor.visitClearCommand(AbstractLockingInterceptor.java:73)
	at org.infinispan.commands.write.ClearCommand.acceptVisitor(ClearCommand.java:33)
	at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNextAndHandle(BaseAsyncInterceptor.java:187)
	at org.infinispan.statetransfer.StateTransferInterceptor.handleNonTxWriteCommand(StateTransferInterceptor.java:312)
	at org.infinispan.statetransfer.StateTransferInterceptor.handleWriteCommand(StateTransferInterceptor.java:256)
	at org.infinispan.statetransfer.StateTransferInterceptor.visitClearCommand(StateTransferInterceptor.java:135)
	at org.infinispan.commands.write.ClearCommand.acceptVisitor(ClearCommand.java:33)
	at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNext(BaseAsyncInterceptor.java:58)
	at org.infinispan.interceptors.DDAsyncInterceptor.handleDefault(DDAsyncInterceptor.java:54)
	at org.infinispan.interceptors.DDAsyncInterceptor.visitClearCommand(DDAsyncInterceptor.java:85)
	at org.infinispan.commands.write.ClearCommand.acceptVisitor(ClearCommand.java:33)
	at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNext(BaseAsyncInterceptor.java:58)
	at org.infinispan.interceptors.DDAsyncInterceptor.handleDefault(DDAsyncInterceptor.java:54)
	at org.infinispan.interceptors.DDAsyncInterceptor.visitClearCommand(DDAsyncInterceptor.java:85)
	at org.infinispan.commands.write.ClearCommand.acceptVisitor(ClearCommand.java:33)
	at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNextAndExceptionally(BaseAsyncInterceptor.java:127)
	at org.infinispan.interceptors.impl.InvocationContextInterceptor.visitCommand(InvocationContextInterceptor.java:90)
	at org.infinispan.interceptors.impl.AsyncInterceptorChainImpl.invokeAsync(AsyncInterceptorChainImpl.java:211)
	at org.infinispan.cache.impl.InvocationHelper.doInvokeAsync(InvocationHelper.java:324)
	at org.infinispan.cache.impl.InvocationHelper.invokeAsync(InvocationHelper.java:157)
	at org.infinispan.cache.impl.CacheImpl.clearAsync(CacheImpl.java:1478)
	at org.infinispan.cache.impl.DecoratedCache.clearAsync(DecoratedCache.java:366)
	at org.infinispan.cache.impl.AbstractDelegatingCache.clearAsync(AbstractDelegatingCache.java:209)
	at org.infinispan.rest.resources.BaseCacheResource.clearEntireCache(BaseCacheResource.java:158)
	at org.infinispan.rest.framework.impl.RestDispatcherImpl.dispatch(RestDispatcherImpl.java:85)
	at org.infinispan.rest.RestRequestHandler.handleRestRequest(RestRequestHandler.java:208)
	at org.infinispan.rest.RestRequestHandler.channelRead0(RestRequestHandler.java:158)
	at org.infinispan.rest.RestRequestHandler.channelRead0(RestRequestHandler.java:54)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at io.netty.handler.codec.http.HttpServerKeepAliveHandler.channelRead(HttpServerKeepAliveHandler.java:64)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at org.infinispan.rest.CorsHandler.channelRead(CorsHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at org.infinispan.rest.StreamCorrelatorHandler.channelRead(StreamCorrelatorHandler.java:27)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:107)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:107)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:107)
	at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:120)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:107)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at org.infinispan.server.core.transport.StatsChannelHandler.channelRead(StatsChannelHandler.java:28)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:799)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:501)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:399)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Infinispan version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java
+++ b/query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java
@@ -4,6 +4,7 @@ import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentMap;
@@ -50,7 +51,7 @@ import org.infinispan.factories.annotations.Inject;
 import org.infinispan.factories.annotations.Start;
 import org.infinispan.factories.annotations.Stop;
 import org.infinispan.interceptors.DDAsyncInterceptor;
-import org.infinispan.interceptors.InvocationSuccessAction;
+import org.infinispan.interceptors.InvocationSuccessFunction;
 import org.infinispan.persistence.manager.PersistenceManager;
 import org.infinispan.persistence.manager.PersistenceManager.StoreChangeListener;
 import org.infinispan.query.core.impl.Log;
@@ -105,7 +106,7 @@ public final class QueryInterceptor extends DDAsyncInterceptor {
    private final DataConversion keyDataConversion;
    private volatile boolean isPersistenceEnabled;
 
-   private final InvocationSuccessAction<ClearCommand> processClearCommand = this::processClearCommand;
+   private final InvocationSuccessFunction<ClearCommand> processClearCommand = this::processClearCommand;
    private final boolean isManualIndexing;
    private final AdvancedCache<?, ?> cache;
    private final Map<String, Class<?>> indexedClasses;
@@ -296,7 +297,7 @@ public final class QueryInterceptor extends DDAsyncInterceptor {
 
    @Override
    public Object visitClearCommand(InvocationContext ctx, ClearCommand command) {
-      return invokeNextThenAccept(ctx, command, processClearCommand);
+      return invokeNextThenApply(ctx, command, processClearCommand);
    }
 
    @Override
@@ -342,13 +343,12 @@ public final class QueryInterceptor extends DDAsyncInterceptor {
    /**
     * Remove all entries from all known indexes
     */
-   public void purgeAllIndexes() {
-      if (searchMapping == null) {
-         return;
+   public Object purgeAllIndexes(InvocationContext ctx, VisitableCommand command, Object rv) {
+      Optional<SearchWorkspace> op;
+      if (searchMapping != null && (op = searchMapping.findScopeAll().map(SearchScope::workspace)).isPresent()) {
+         return asyncInvokeNext(ctx, command, op.get().purgeAsync());
       }
-
-      // avoid starting if not created yet
-      searchMapping.findScopeAll().map(SearchScope::workspace).ifPresent(SearchWorkspace::purge);
+      return rv;
    }
 
    public void purgeIndex(Class<?> entityType) {
@@ -468,10 +468,11 @@ public final class QueryInterceptor extends DDAsyncInterceptor {
       return value != null && previousValue != null && value.getClass() != previousValue.getClass();
    }
 
-   private void processClearCommand(InvocationContext ctx, ClearCommand command, Object rv) {
+   private Object processClearCommand(InvocationContext ctx, ClearCommand command, Object rv) {
       if (shouldModifyIndexes(command, ctx, null)) {
-         purgeAllIndexes();
+         return purgeAllIndexes(ctx, command, rv);
       }
+      return rv;
    }
 
    public boolean isStopping() {

--- a/query/src/main/java/org/infinispan/search/mapper/scope/SearchWorkspace.java
+++ b/query/src/main/java/org/infinispan/search/mapper/scope/SearchWorkspace.java
@@ -1,6 +1,7 @@
 package org.infinispan.search.mapper.scope;
 
 import java.util.Set;
+import java.util.concurrent.CompletionStage;
 
 /**
  * The entry point for explicit index operations.
@@ -29,6 +30,12 @@ public interface SearchWorkspace {
     * If empty, documents will be deleted regardless of their routing key.
     */
    void purge(Set<String> routingKeys);
+
+   /**
+    * Asynchronous version of {@link #purge()}
+    * @return stage that when complete the purge will be complete
+    */
+   CompletionStage<?> purgeAsync();
 
    /**
     * Flush to disk the changes to indexes that were not committed yet. In the case of backends with a transaction log

--- a/query/src/main/java/org/infinispan/search/mapper/scope/impl/SearchWorkspaceImpl.java
+++ b/query/src/main/java/org/infinispan/search/mapper/scope/impl/SearchWorkspaceImpl.java
@@ -2,6 +2,7 @@ package org.infinispan.search.mapper.scope.impl;
 
 import java.util.Collections;
 import java.util.Set;
+import java.util.concurrent.CompletionStage;
 
 import org.hibernate.search.mapper.pojo.work.spi.PojoScopeWorkspace;
 import org.hibernate.search.util.common.impl.Futures;
@@ -16,7 +17,12 @@ class SearchWorkspaceImpl implements SearchWorkspace {
 
    @Override
    public void purge() {
-      Futures.unwrappedExceptionJoin(delegate.purge(Collections.emptySet()));
+      Futures.unwrappedExceptionJoin(purgeAsync().toCompletableFuture());
+   }
+
+   @Override
+   public CompletionStage<?> purgeAsync() {
+      return delegate.purge(Collections.emptySet());
    }
 
    @Override

